riscv: translate virtual address to physical address.

Hsiangkai opened the pull request:
Add RISC-V address translation algorithm for SV32, SV39, and SV48. We have tested the patch on HiFive Unleashed platform and gdb could read virtual memory successfully.

Hsiangkai committed:
riscv: translate virtual address to physical address.

Hsiangkai committed:
riscv: fix formatting errors.

Hsiangkai committed:
riscv: fix build errors.

timsifive commented on src/target/riscv/riscv.c:
```diff
@@ -2386,6 +2594,13 @@ static const struct command_registration riscv_exec_command_handlers[] = {
 			"1: DATA_REGISTER}"
 	},
+	{
+		.name = "disable_virt2phys",
```
Can you change this from disable to enable, and have it enabled by default? That avoids commands with double negatives like "riscv disable_virt2phy off" which can be hard to understand.

Please also document this new command in doc/openocd.texi.

Hsiangkai replied:
I found there is already a command called 'set_enable_virtual' in the document. I will reuse the command to turn on/off the virtual address translation.

timsifive reviewed (changes requested):
Looks great. Thank you!

I have just one more request.

Hsiangkai committed:
riscv: Remove redundant command for virtual address access.

timsifive commented:
That command already controls address translation by setting dcsr.mprven, which is a pretty different method from what you implemented. Ideally there is one single command that gives the following address translation options:
1. off
2. mprven, what is currently controlled by riscv_enable_virtual,
3. translate, what you implemented

I had planned to do that cleanup later. But regardless, we can't use the same command for both 2 and 3, which is what's happening with your latest change.

Hsiangkai committed:
Revert "riscv: Remove redundant command for virtual address access."

This reverts commit 990d09eac37d2effcfc5c0d0b5c99678f45e7d7f.

Hsiangkai committed:
riscv: Change command disable_virt2phys  to set_enable_virt2phys

1. Avoid double negative logic to make users easy to use.
2. Add document about new comomand 'riscv set_enable_virt2phys on|off'

timsifive reviewed (approved):
Thank you!

paul-walmsley-sifive commented:
Very excited for this to land.  Any projections on when it might be merged?

timsifive commented:
@paul-walmsley-sifive, I usually merge the day after I approve, in case there's some last-minute feedback.

paul-walmsley-sifive commented:
Cool!  Thanks!
